File: tests/fixtures/code/sample.go
// Package search is a small client for a built index, invented for the test suite.
package search

import (
	"fmt"
	"strings"
)

// Limit is the most results one search returns.
const Limit = 10

// Result is one ranked result.
type Result struct {
	ParentID string
	Score    float64
}

// Ranker orders results.
type Ranker interface {
	Rank(results []Result) []Result
}

// Compendium reads a built index and answers questions about it.
type Compendium struct {
	titles []string
}

// Search returns the titles that hold the term.
func (c *Compendium) Search(term string) []string {
	var found []string
	for _, title := range c.titles {
		if strings.Contains(title, term) {
			found = append(found, title)
		}
	}
	return found
}

// Describe prints how many titles the compendium holds.
func Describe(c *Compendium) {
	fmt.Println(len(c.titles))
}

File: tests/fixtures/code/sample.rs
//! A small client for a built index, invented for the test suite.

use std::collections::HashMap;

/// The most results one search returns.
const LIMIT: usize = 10;

/// One ranked result.
pub struct Result {
    pub parent_id: String,
    pub score: f64,
}

/// Orders results.
pub trait Ranking {
    fn rank(&self, results: Vec<Result>) -> Vec<Result>;
}

/// Reads a built index and answers questions about it.
pub struct Compendium {
    titles: Vec<String>,
}

impl Compendium {
    /// Returns the titles that hold the term.
    pub fn search(&self, term: &str) -> Vec<String> {
        self.titles.iter().filter(|t| t.contains(term)).cloned().collect()
    }
}

pub mod util {
    /// Counts how often each title appears.
    pub fn tally(titles: &[String]) -> super::HashMap<String, usize> {
        let mut counts = super::HashMap::new();
        for title in titles {
            *counts.entry(title.clone()).or_insert(0) += 1;
        }
        counts
    }
}

/// Prints how many titles the compendium holds.
pub fn describe(compendium: &Compendium) {
    println!("{}", compendium.titles.len());
}
